servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -316,115 +410,189 @@ public Completable closeAsyncGracefully() {
     }
 
     /**
-     * {@link LoadBalancerFactory} for {@link RoundRobinLoadBalancer}.
+     * Please use {@link io.servicetalk.loadbalancer.RoundRobinLoadBalancerFactory} instead of this factory.
      *
      * @param <ResolvedAddress> The resolved address type.
      * @param <C> The type of connection.
+     * @deprecated Use {@link io.servicetalk.loadbalancer.RoundRobinLoadBalancerFactory} to build instances
+     * of this {@link LoadBalancer}
      */
+    @Deprecated
     public static final class RoundRobinLoadBalancerFactory<ResolvedAddress, C extends LoadBalancedConnection>
             implements LoadBalancerFactory<ResolvedAddress, C> {
 
         @Override
         public <T extends C> LoadBalancer<T> newLoadBalancer(
                 final Publisher<? extends ServiceDiscovererEvent<ResolvedAddress>> eventPublisher,
                 final ConnectionFactory<ResolvedAddress, T> connectionFactory) {
-            return new RoundRobinLoadBalancer<>(eventPublisher, connectionFactory);
+            return new RoundRobinLoadBalancer<>(eventPublisher, connectionFactory, EAGER_CONNECTION_SHUTDOWN_ENABLED);
         }
     }
 
     // Visible for testing
-    List<Entry<ResolvedAddress, List<C>>> activeAddresses() {
-        return activeHosts.stream().map(Host::asEntry).collect(toList());
+    List<Entry<ResolvedAddress, List<C>>> usedAddresses() {
+        return usedHosts.stream().map(Host::asEntry).collect(toList());
     }
 
-    private static final class Host<Addr, C extends ListenableAsyncCloseable> implements AsyncCloseable {
-        @SuppressWarnings("rawtypes")
-        private static final AtomicReferenceFieldUpdater<Host, Object[]> connectionsUpdater =
-                AtomicReferenceFieldUpdater.newUpdater(Host.class, Object[].class, "connections");
+    private static final class Host<Addr, C extends ListenableAsyncCloseable> implements ListenableAsyncCloseable {
+        private static final ConnState EMPTY_CONN_STATE = new ConnState(EMPTY_ARRAY, State.ACTIVE);
+        private static final ConnState CLOSED_CONN_STATE = new ConnState(EMPTY_ARRAY, State.CLOSED);
+
+        private enum State {
+            ACTIVE,
+            EXPIRED,
+            CLOSED
+        }
+
+        private static final class ConnState {
+            final Object[] connections;
+            final State state;
+
+            ConnState(final Object[] connections, final State state) {
+                this.connections = connections;
+                this.state = state;
+            }
+        }
 
         final Addr address;
-        private volatile Object[] connections = EMPTY_ARRAY;
+        private final ListenableAsyncCloseable closeable;
+
+        @SuppressWarnings("rawtypes")
```

Comment:
I thought we had quality rules for this but typically the `private static final` methods are all grouped together at the top of the class, and non-static state is below that.

Reply:
Unfortunately, check style does not have this type of validation (to my knowledge).

Reply:
We have https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/DeclarationOrderCheck.html configured with `ignoreModifiers=true`. The check follows the original Sun Java Code Conventions document and according to it all methods go below class/instance fields.

Reply:
Our checkstyle configuration uses https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/coding/DeclarationOrderCheck.html with `<property name="ignoreModifiers" value="true"/>`, which disables such detection. Should we re-enable it? When that property is not set, the order is properly verified.

Reply:
@chemicL - lets consider check style updates in a followup/isolated PR, but yes this seems interesting if it leads to more standardization.